File: leetcode/javascript/000023.merge-k-sorted-lists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */

/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 * Time complexity: O(nlog(n))
 * Space complexity: O(n)
 */
var mergeKLists = function (lists) {
  const values = [];
  for (let list of lists) {
    while (list) {
      values.push(list.val);
      list = list.next;
    }
  }

  values.sort((a, b) => a - b);

  let dummyHead = new ListNode(0);
  let cur = dummyHead;
  for (const val of values) {
    cur.next = new ListNode(val);
    cur = cur.next;
  }

  return dummyHead.next;
};

/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 * Time complexity: O(nlog(k))
 * Space complexity: O(k)
 */
var mergeKLists = function (lists) {
  if (lists.length === 0) return null;
  const minHeap = new MinPriorityQueue((x) => x.val);
  for (let list of lists) {
    if (list !== null) minHeap.enqueue(list);
  }

  let dummyHead = new ListNode(0);
  let cur = dummyHead;
  while (minHeap.size() > 0) {
    let node = minHeap.dequeue();
    cur.next = node;
    cur = cur.next;

    node = node.next;
    if (node) minHeap.enqueue(node);
  }

  return dummyHead.next;
};

File: leetcode/python/000023.merge-k-sorted-lists.py
# Definition for singly-linked list.
import heapq
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class NodeWrapper:
    def __init__(self, node):
        self.node = node

    def __lt__(self, other):
        return self.node.val < other.node.val


class Solution:
    # Time complexity: O(nlog(k))
    # Space complexity: O(k)
    def mergeKLists(self, lists: list[Optional[ListNode]]) -> Optional[ListNode]:
        if len(lists) == 0:
            return None

        res = ListNode(0)
        cur = res
        min_heap = []

        for l in lists:
            if l:
                heapq.heappush(min_heap, NodeWrapper(l))

        while min_heap:
            node_wrapper = heapq.heappop(min_heap)
            cur.next = node_wrapper.node
            cur = cur.next

            if node_wrapper.node.next:
                heapq.heappush(min_heap, NodeWrapper(node_wrapper.node.next))

        return res.next
